xapian-core/include/xapian/weight.h

```diff
@@ -409,6 +409,11 @@ class XAPIAN_VISIBILITY_DEFAULT TfIdfWeight : public Weight {
     /// The factor to multiply with the weight.
     double factor;
 
+    /// Parameters slope and delta in the Piv+ normalization weighting formula.
+    /// Default values: 0.2 and 1.0 respectively.
+    double param_slope = 0.2;
+    double param_delta = 1.0;
+
     TfIdfWeight * clone() const;
```

Comment:
My reading of the paper is that they are saying they tuned the parameters within the ranges they give in each case, but that doesn't mean that any values outside those ranges are invalid (just that the authors think the optimal values are likely to be within those ranges, for reasons they don't seem to explain).

Here they describe the tuning for BM25 parameter's `b` and `k1`:
> In both BM25+ and BM25, we train b and k1 using cross validation, where b is tuned from 0.1 to 0.9 in increments of 0.1, and k1 is tuned from 0.2 to 4.0 in increments of 0.2.

But BM25 only requires `0<=b<=1` and `k1>=0`.

Some values of the parameters added in the + variants may actually be invalid (I don't see anything explicitly saying it, but it seems in each case δ should be positive, or at least non-negative) - I think we should reject any values which are really invalid, but still allow values outside the ranges Lv and Zhai tuned within as these may still be interesting.

Checking the paper, I agree that they don't really suggest default values for Piv+ (table 10 suggests values for `s` between 0.01 and 0.19 depending on the collection and nature of the query, which is a sufficiently large range that picking a single good default is hard).

Rejecting `P` for `normalizations[2]` in the default constructor so the user has to specify slope and delta seems a good plan - I'm not sure we need to reject `P` for the other two normalisations though - they don't seem to use slope or delta.

I think in the "full" constructor that `slope <= 0` or `delta <= 0` should be rejected with an exception - that's what we do for most of the weighting schemes.  `BM25Weight` just clamps parameters to be within the valid range because it always has and people could be relying on that happening, but for the schemes implemented more recently we decided rejecting invalid values with an exception was more helpful to users.  And as you note, if we force them to something valid we have to pick a value to force them to, which would just bring us back to effectively picking a good default.